src/net.cpp

```diff
@@ -754,6 +766,18 @@ int CNetMessage::readData(const char *pch, unsigned int nBytes)
     memcpy(&vRecv[nDataPos], pch, nCopy);
     nDataPos += nCopy;
 
+    if (complete())
+    {
+        // reject messages with invalid checksum
+        uint256 hash = Hash(vRecv.begin(), vRecv.begin() + hdr.nMessageSize);
+        unsigned int nChecksum = ReadLE32((unsigned char*)&hash);
+        if (nChecksum != hdr.nChecksum)
+        {
+            LogPrint("net" "Message error: invalid checksum nChecksum=%08x hdr.nChecksum=%08x\n", __func__, nChecksum, hdr.nChecksum);
+            return -1;
+        }
+    }
+
```

Comment:
I have a bit of concern about moving the checksum evaluation from `ProcessMessages` to `CNetMessage::readData` function, as it implicitly moves the execution context of this check from `threadMessageHandler` to `threadContextHandler`.
As now it does not make much difference, as the presence of the `cs_vRecvMsg` lock guarantees that only one of the two threads is running, but in the future we plan to disjoint the two threads. This will imply that zend will read the socket and process the messages (almost) asynchronously; for this reason, I suppose that `threadMessageHandler` should be as quick as possible in reading the socket, leveraging all the heavy checks to the message processing thread.
Unfortunately, it's very hard to forecast how leaving this check here would affect performance, but I am open to discussion whether leave it as it is for the moment, and make a change only after the final refactoring.